### PR TITLE
Implement graceful shutdown service dependencies

### PR DESCRIPTION
**Pull request by tso**
```13:09:55.223 [JettyService STARTING] INFO misk.web.jetty.JettyService -- Started Jetty in 280.7 ms on port 0/0
13:09:56.265 [JettyService STOPPING] INFO misk.web.jetty.JettyService -- Stopping Jetty
13:09:56.266 [JettyService STOPPING] INFO org.eclipse.jetty.server.Server -- Stopped Server@2175b71{STOPPING}[10.0.14,sto=25000]
13:09:56.267 [JettyService STOPPING] INFO org.eclipse.jetty.server.Server -- Shutdown Server@2175b71{STOPPING}[10.0.14,sto=25000]
13:10:01.346 [JettyService STOPPING] INFO org.eclipse.jetty.server.AbstractConnector -- Stopped health@4be8cb9f{HTTP/1.1, (http/1.1)}{0.0.0.0:0}
13:10:01.347 [JettyService STOPPING] INFO org.eclipse.jetty.server.AbstractConnector -- Stopped http@6156232d{HTTP/1.1, (http/1.1, h2c)}{127.0.0.1:0}
13:10:01.354 [JettyService STOPPING] INFO org.eclipse.jetty.server.AbstractConnector -- Stopped https@56ad378e{SSL, (ssl, alpn, h2, http/1.1)}{127.0.0.1:0}
13:10:01.357 [JettyService STOPPING] INFO org.eclipse.jetty.server.handler.ContextHandler -- Stopped o.e.j.s.ServletContextHandler@2e70ed5f{/,null,STOPPED}
13:10:01.367 [JettyService STOPPING] INFO misk.web.jetty.JettyService -- Stopped Jetty in 5.102 s
13:10:01.368 [ReadyService STOPPING] INFO misk.ReadyService -- Shutting down
13:10:01.368 [FakeService STOPPING] INFO misk.web.FakeService -- Shutting down
```

from the test i added. Note that jetty takes 5 seconds to shutdown since there's an HTTP call in flight

### CODE REVIEW

File: misk-service/src/main/kotlin/misk/ReadyService.kt [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+ *
+ * Thus, we can force JettyService (or SQS, or tasks, or kinesis) to stop only once all the
+ * core infra services have stopped!
+ *
+ * Services that need everything to be ready other than themselves should depend on ReadyService
+ *
+ * Services that are more base should be enhancedBy ReadyService
+ */
+@Singleton
+class ReadyService @Inject constructor() : AbstractIdleService() {
```
Comment (tso) [inline misk-service/src/main/kotlin/misk/ReadyService.kt:19]: Different name perhaps?

File: misk/src/test/kotlin/misk/web/GracefulShutdownTest.kt [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+  @AfterEach
+  fun tearDown() {
+    httpClient.dispatcher.executorService.shutdown()
+  }
+
+  @Test
+  fun basic() {
+    makeHttpCall()
+    Thread.sleep(1000)
```
Comment (tso) [inline misk/src/test/kotlin/misk/web/GracefulShutdownTest.kt:60]: this feels hacky -- how to share state between this test and `HelloAction`?

File: misk/src/test/kotlin/misk/web/GracefulShutdownTest.kt [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+import wisp.logging.getLogger
+import java.io.IOException
+import java.time.Duration
+import java.util.concurrent.CountDownLatch
+import java.util.concurrent.TimeUnit
+import javax.inject.Inject
+import javax.inject.Singleton
+
+@MiskTest(startService = true)
+internal class GracefulShutdownTest {
```
Comment (jdm-square) [inline misk/src/test/kotlin/misk/web/GracefulShutdownTest.kt:32]: Could you add documentation for what this test is attempting to validate and how it's doing that?

**Review by jdm-square — COMMENTED** [review]
I like this look of this so far. I don't have a good answer for how we ensure that any changes to misk appropriately use dependsUpon/enhancedBy with ReadyService, but it looks better than the status quo!

**Comment by tso** [comment]
> I like this look of this so far. I don't have a good answer for how we ensure that any changes to misk appropriately use dependsUpon/enhancedBy with ReadyService, but it looks better than the status quo!

Yeah I don't either but these types of services are added rarely i think (misk itself has probably added 0 in the last 3 years?)

File: misk-service/src/main/kotlin/misk/ReadyService.kt (hunk 1 above)
Comment (swankjesse) [reply misk-service/src/main/kotlin/misk/ReadyService.kt:21]: I think the docs need more . . . explanation of this as a symbolic service, useful to depend on or have depend on you?

**Review by swankjesse — APPROVED** [review]

File: misk-service/src/main/kotlin/misk/ReadyService.kt (hunk 1 above)
Comment (tso) [reply misk-service/src/main/kotlin/misk/ReadyService.kt:21]: Yeah, you're definitely right.

How's 
```
/**
 * This is a symbolic service that's useful to define the relationship, generally, between
 * services which process traffic (Jetty, SQS, Kinesis, Cron, Tasks, etc.) and services which
 * are required to do work (Database, Redis, GCP, Feature Flags).
 *
 * By having the former depend on ReadyService and the latter enhance ReadyService we can force,
 * for example, JettyService to stop _before_ our feature flag service without having to intertwine
 * our dependency graph.
 */
```

Comment (jdm-square) [reply misk-service/src/main/kotlin/misk/ReadyService.kt:21]: That's a nice improvement.

File: misk-cron/src/main/kotlin/misk/cron/CronModule.kt [hunk 4]
```diff
@@ -20,12 +21,12 @@ class CronModule(
 ) : KAbstractModule() {
   override fun configure() {
     install(FakeCronModule(zoneId, threadPoolSize, dependencies))
-    install(ServiceModule<RepeatedTaskQueue>(ForMiskCron::class))
+    install(ServiceModule<RepeatedTaskQueue>(ForMiskCron::class).dependsOn<ReadyService>())
```
Comment (tso) [inline misk-cron/src/main/kotlin/misk/cron/CronModule.kt:24]: This is backwards incompatible -- Need to update any users of this who specify dependencies that ReadyService enhances

**Review by ean5533 — APPROVED** [review]
Love it.

**Review by ean5533 — APPROVED** [review]

File: misk-service/src/test/kotlin/misk/ServiceGraphBuilderTest.kt [hunk 5]
```diff
@@ -49,8 +49,8 @@ class ServiceGraphBuilderTest {
     assertThat(script).isEqualTo(
       """
         |starting Service A
-        |starting Enhancement A
         |starting Service C
+        |starting Enhancement A
```
Comment (keeferrourke) [inline misk-service/src/test/kotlin/misk/ServiceGraphBuilderTest.kt:53]: Is this not a significant behaviour change? Enhancements of a service are meant to start before dependents of the service.

Comment (tso) [reply misk-service/src/test/kotlin/misk/ServiceGraphBuilderTest.kt:53]: oh interesting, good catch. I think it's fine in this case though given it's just `SchemaMigratorService`

Comment (keeferrourke) [reply misk-service/src/test/kotlin/misk/ServiceGraphBuilderTest.kt:53]: Maybe we ought to just rip out service enhancements entirely and add a `.dependsUpon` directive to allow expressing the inverse dep relationship that "enhancements" are now expressing 🤔

**Pull request merged**